viola/src/libIMG/xloadimage.man
.TH XLOADIMAGE 1 "8 May 1991"
.SH NAME
xloadimage, xsetbg, xview \- load images into an X11 window or onto
the root window
.SH SYNOPSIS
\fIxloadimage\fR [global_options] {[image_options] image ...}
.br
\fIxloadimage\fR [global_options] [image_options] stdin < image
.SH DESCRIPTION
\fIXloadimage\fR displays images in an X11 window or loads them onto
the root window.  See the \fIIMAGE TYPES\fR section below for
supported image types.
.PP
If the filename \fIstdin\fR is given, xloadimage will read the image
from standard input.
.PP
If the destination display cannot support the number of colors in the
image, the image will be dithered (monochrome destination) or have its
colormap reduced (color destination) as appropriate.  This can also be
done forcibly with the \fI-halftone\fR, \fI-dither\fR, and
\fI-colors\fR options.
.PP
A variety of image manipulations can be specified, including gamma
correction, brightening, clipping, dithering, depth-reduction,
rotation, and zooming.  Most of these manipulations have simple
implementations; speed was opted for above accuracy.
.PP
If you are viewing a large image in a window, the initial window will
be at most 90% of the size of the display unless the window manager
does not correctly handle window size requests or if you've used the
\fI-fullscreen\fR option.  You may move the image around in the window
by dragging with the first mouse button.  The cursor will indicate
which directions you may drag, if any.  You may exit the window by
typing 'q' or '^C' when the keyboard focus is on the window.
.PP
It's possible to have a "slideshow" of many images by specifying the
\fI-slideshow\fR option.
.PP
A wide variety of common image manipulations can be done by mixing and
matching the available options.  See the section entitled \fIHINTS FOR
GOOD IMAGE DISPLAYS\fR for some ideas.
.PP
\fIXsetbg\fR is equivalent to \fIxloadimage -onroot -quiet\fR and
\fIxview\fR is equivalent to \fIxloadimage -view -verbose\fR.
.SH RESOURCE CLASS
\fIXloadimage\fR uses the resource class name \fIXloadimage\fR for
window managers which need this resource set.  This name changed
in version 2.00 and 2.01; some previous versions used the name
\fIXLoadImage\fR (which was difficult to predict) or \fIxloadimage\fR
(which conflicted with class naming conventions).
.SH GLOBAL OPTIONS
The following options affect the global operation of \fIxloadimage\fR.
They may be specified anywhere on the command line.
.TP 8
-border \fIcolor\fR
This sets the background portion of the window which is not covered by
any images to be \fIcolor\fR.
.TP
-default
Set the root background to the default root weave.  This is the same
as \fIxsetroot\fR with no arguments.
.TP
-debug
Talk to the X server in synchronous mode.  This is useful for
debugging.  If an X error is seen while in this mode, a core will be
dumped.
.TP
-delay \fIsecs\fR
Automatically advance to the next image after \fIsecs\fR seconds.
.TP
-display \fIdisplay_name\fR
X11 display name to send the image(s) to.
.TP
-fit
Force image to use the default visual and colormap.  This is useful if
you do not want technicolor effects when the colormap focus is inside
the image window, but it may reduce the quality of the displayed
image.  This is on by default if -onroot or -windowid is specified.
.TP
-fork
Fork xloadimage.  This causes xloadimage to disassociate itself from
the shell.  This option automatically turns on -quiet.
.TP
-fullscreen
Use the entire screen to display images.  If combined with -onroot,
the image will be zoomed to fill the entire rootwindow.
.TP
-geometry \fIWxH[{+-X}{+-}Y]\fR
This sets the size of the window onto which the images are loaded to a
different value than the size of the image.  When viewing an image in
a window, this can be used to reduce the size of the destination
window.  When loading an image onto the root window, this option
controls the size of the pixmap which will be loaded onto the root.
If the size is smaller than that of the display, the image will be
replicated.
.TP
-goto image_name
Forces the next image to be displayed to be the image named
\fIimage_name\fR.  This is useful for generating looped slideshows.
If more than one image of the same name as the target exists on the
argument list, the first in the argument list is used.
.TP
-help [option ...]
Give information on an option or list of options.  If no option is
given, a simple interactive help facility is invoked.
.TP
-identify
Identify the supplied images rather than display them.
.TP
-install
Forcibly install the image's colormap when the window is focused.
This violates ICCCM standards and only exists to allow operation with
naive window managers.  Use this option only if your window manager
does not install colormaps properly.
.TP
-list
List the images which are along the image path.
.TP
-onroot
Load image(s) onto the root window instead of viewing in a window.
This option automatically sets the -fit option.
This is the opposite of \fI-view\fR.  \fIXSetbg\fR has this option set
by default.
.TP
-path
Displays the image path and image suffixes which will be used when
looking for images.  These are loaded from ~/.xloadimagerc and
optionally from a systemwide file (normally /usr/lib/xloadimagerc).
.TP
-pixmap
Force the use of a pixmap as backing-store.  This is provided for
servers where backing-store is broken (such as some versions of the
AIXWindows server).  It may improve scrolling performance on servers
which provide backing-store.
.TP
-private
Force the use of a private colormap.  Normally colors are allocated
shared unless there are not enough colors available.
.TP
-quiet
Forces \fIxloadimage\fR and \fIxview\fR to be quiet.  This is the
default for \fIxsetbg\fR, but the others like to whistle. 
.TP
-supported
List the supported image types. 
.TP
-verbose
Causes \fIxloadimage\fR to be talkative, telling you what kind of
image it's playing with and any special processing that it has to do. 
This is the default for \fIxview\fR and \fIxloadimage\fR. 
.TP
-version
Print the version number and patchlevel of this version of
\fIxloadimage\fR.
.TP
-view
View image(s) in a window.  This is the opposite of \fI-onroot\fR and
the default for \fIxview\fR and \fIxloadimage\fR. 
.TP
-visual \fIvisual_name\fR
Force the use of a specific visual type to display an image.  Normally
\fIxloadimage\fR tries to pick the best available image for a
particular image type.  The available visual types are:  DirectColor,
TrueColor, PseudoColor, StaticColor, GrayScale, and StaticGray.
Nonconflicting names may be abbreviated and case is ignored.
.TP
-windowid \fIhex_window_id\fR
Sets the background pixmap of a particular window ID.  The argument
must be in hexadecimal and must be preceeded by "0x" (\fIeg\fR
-windowid 0x40000b.  This is intended for setting the background
pixmap of some servers which use untagged virtual roots
(\fIeg\fR HP-VUE), but can have other interesting applications.
.SH IMAGE OPTIONS
The following options may preceed each image.  These options are
local to the image they preceed. 
.TP
-at \fIX\fR,\fIY\fR
Indicates coordinates to load the image at on the base image.  If
this is an option to the first image, and the \fI-onroot\fR option is
specified, the image will be loaded at the given location on the
display background. 
.TP
-background \fIcolor\fR
Use \fIcolor\fR as the background color instead of the default
(usually white but this depends on the image type) if you are
transferring a monochrome image to a color display. 
.TP
-brighten \fIpercentage\fR
Specify a percentage multiplier for a color image's colormap.  A value
of more than 100 will brighten an image, one of less than 100 will
darken it. 
.TP
-center
Center the image on the base image loaded.  If this is an option to
the first image, and the \fI-onroot\fR option is specified, the image
will be centered on the display background. 
.TP
-clip \fIX\fR,\fIY\fR,\fIW\fR,\fIH\fR
Clip the image before loading it.  \fIX\fR and \fIY\fR define the
upper-left corner of the clip area, and \fIW\fR and \fIH\fR define the
extents of the area.  A zero value for \fIW\fR or \fIH\fR will be
interpreted as the remainder of the image. 
.TP
-colors \fIn\fR
Specify the maximum number of colors to use in the image.  This is a
way to forcibly reduce the depth of an image.
.TP
-dither
Dither a color image to monochrome using a Floyd-Steinberg dithering
algorithm.  This happens by default when viewing color images on a
monochrome display.  This is slower than \fI-halftone\fR and affects
the image accuracy but usually looks much better.
.TP
-foreground \fIcolor\fR
Use \fIcolor\fR as the foreground color instead of black if you are
transferring a monochrome image to a color display.  This can also be
used to invert the foreground and background colors of a monochrome
image. 
.TP
-gamma \fIdisplay_gamma\fR
Specify the gamma correction for the display.
The default value is 1.0, a typical display needs 2.0 to 2.5.
.TP
-gray
Convert an image to grayscale.  This is very useful when displaying
colorful images on servers with limited color capability.  The
optional spelling \fI-grey\fR may also be used.
.TP
-halftone
Force halftone dithering of a color image when displaying on a
monochrome display.  This option is ignored on monochrome images.
This dithering algorithm blows an image up by sixteen times; if you
don't like this, the \fI-dither\fR option will not blow the image up
but will take longer to process and will be less accurate.
.TP
-idelay \fIsecs\fR
Set the delay to be used for this image to \fIsecs\fR seconds (see
\fI-delay\fR).  If \fI-delay\fR was specified, this overrides it.  If
it was not specified, this sets the automatic advance delay for this
image while others will wait for the user to advance them.
.TP
-invert
Inverts a monochrome image.  This is shorthand for \fI-foreground
white -background black\fR.
.TP
-merge
Merge this image onto the base image after local processing.  The base
image is considered to be the first image specified or the last image
that was not preceeded by \fI-merge\fR.  If used in conjunction with
\fI-at\fR and \fI-clip\fR, very complex images can be built up.  This
option is on by default for all images if the \fI-onroot\fR or
\fI-windowid\fR options are specified.
.TP
-name \fIimage_name\fR
Force the next argument to be treated as an image name.  This is
useful if the name of the image is \fI-dither\fR, for instance. 
.TP
-newoptions
Reset options that propagate.  The \fI-bright, -colors, -delay,
-dither, -gamma, -normalize, -smooth, -xzoom, -yzoom\fR, and
\fI-zoom\fR options normally propagate to all following images.
.TP
-normalize
Normalize a color image.
.TP
-rotate \fIdegrees\fR
Rotate the image by \fIdegrees\fR clockwise.  The number must be a
multiple of 90.
.TP
-smooth
Smooth a color image.  This reduces blockiness after zooming an image
up.  If used on a monochrome image, nothing happens.  This option can
take awhile to perform, especially on large images.  You may specify
more than one \fI-smooth\fR option per image, causing multiple
iterations of the smoothing algorithm.
.TP
-xzoom \fIpercentage\fR
Zoom the X axis of an image by \fIpercentage\fR.  A number greater
than 100 will expand the image, one smaller will compress it.  A zero
value will be ignored.  This option, and the related \fI-yzoom\fR are
useful for correcting the aspect ratio of images to be displayed.
.TP
-yzoom \fIpercentage\fR
Zoom the Y axis of an image by \fIpercentage\fR.  See \fI-xzoom\fR for
more information. 
.TP
-zoom \fIpercentage\fR
Zoom both the X and Y axes by \fIpercentage\fR.  See \fI-xzoom\fR for
more information.  Technically the percentage actually zoomed is the
square of the number supplied since the zoom is to both axes, but I
opted for consistency instead of accuracy.
.SH EXAMPLES
To load the rasterfile "my.image" onto the background and replicate
it to fill the entire background:
.sp
.ti +5
xloadimage -onroot my.image
.PP
To load a monochrome image "my.image" onto the background, using red
as the foreground color, replicate the image, and overlay
"another.image" onto it at coordinate (10,10):
.sp
.ti +5
xloadimage -foreground red my.image -at 10,10 another.image
.PP
To center the rectangular region from 10 to 110 along the X axis and
from 10 to the height of the image along the Y axis:
.sp
.ti +5
xloadimage -center -clip 10,10,100,0 my.image
.PP
To double the size of an image:
.sp
.ti +5
xloadimage -zoom 200 my.image
.PP
To halve the size of an image:
.sp
.ti +5
xloadimage -zoom 50 my.image
.PP
To brighten a dark image:
.sp
.ti +5
xloadimage -brighten 150 my.image
.PP
To darken a bright image:
.sp
.ti +5
xloadimage -brighten 50 my.image
.SH HINTS FOR GOOD IMAGE DISPLAYS
Since images are likely to come from a variety of sources, they may be
in a variety of aspect ratios which may not be supported by your
display.  The \fI-xzoom\fR and \fI-yzoom\fR options can be used to
change the aspect ratio of an image before display.  If you use these
options, it is recommended that you increase the size of one of the
dimensions instead of shrinking the other, since shrinking looses
detail.  For instance, many GIF and G3 FAX images have an X:Y ratio of
about 2:1.  You can correct this for viewing on a 1:1 display with
either \fI-xzoom 50\fR or \fI-yzoom 200\fR (reduce X axis to 50% of
its size and expand Y axis to 200% of its size, respectively) but the
latter should be used so no detail is lost in the conversion.
.PP
When zooming color images up you can reduce blockiness with

viola/src/libIMG/xloadimage.c
/* xloadimage.c:
 *
 * generic image loader for X11
 *
 * jim frost 09.27.89
 *
 * Copyright 1989, 1990, 1991 Jim Frost.
 * See included file "copyright.h" for complete copyright information.
 */

#include "copyright.h"
#include "xloadimage.h"
#include "patchlevel"
#include <signal.h>

/*PYW begin*/
typedef struct ImgObj {
  Image *image;
  XImageInfo *ximageinfo;
  Window w;
} ImgObj;
ImgObj imgObj[10];
int imgObjIdx = 0;
GC gc;
/*PYW end*/

extern double atof();
extern int _Xdebug;

/* if an image loader needs to have our display and screen, it will get
 * them from here.  this is done to keep most of the image routines
 * clean
 */

Display      *Disp= NULL;
int           Scrn= 0;

/* used for the -default option.  this is the root weave bitmap with
 * the bits in the order that xloadimage likes.
 */

#define root_weave_width 4
#define root_weave_height 4
static byte root_weave_bits[] = {
  0xe0, 0xb0, 0xd0, 0x70
};

int           dispcells;
Colormap      theCmap;


/* the real thing
 */

main(argc, argv)
     int argc;
     char *argv[];
{ char         *border;
  char         *dname;
  Display      *disp;           /* display we're sending to */
  int           scrn;           /* screen we're sending to */

  Image        *dispimage;      /* image that will be sent to the display */
  Image        *newimage;       /* new image we're loading */
  Image        *tmpimage;
  XColor        xcolor;         /* color for border option */
  ImageOptions  images[MAXIMAGES + 1]; /* list of image w/ options to load */

  int           a;
  unsigned int  imagecount;     /* number of images in ImageName array */
  char         *user_geometry;	/* -geometry passed by user */
  unsigned int  dest_window;    /* window id to put image onto */
  unsigned int  do_fork;
  unsigned int  fit;
  unsigned int  fullscreen;
  unsigned int  identify;
  unsigned int  install;
  unsigned int  onroot;
  unsigned int  private_cmap;
  unsigned int  set_default;
  unsigned int  use_pixmap;
  unsigned int  verbose;
  int           visual_class;    /* user-defined visual class */
  unsigned int  winwidth, winheight; /* geometry of image */
  unsigned int  global_bright= 0; /* options which are propagated to all */
  unsigned int  global_colors= 0; /* remaining images */
  unsigned int  global_delay= 0;
  unsigned int  global_dither= 0;
  float         global_gamma= 1.0;
  unsigned int  global_normalize= 0;
  unsigned int  global_smooth= 0;
  unsigned int  global_xzoom= 0;
  unsigned int  global_yzoom= 0;

  /* set up internal error handlers
   */

  signal(SIGSEGV, internalError);
  signal(SIGBUS, internalError);
  signal(SIGFPE, internalError);
  signal(SIGILL, internalError);
#if defined(_AIX) && defined(_IBMR2)
  /* the RS/6000 (AIX 3.1) has a new signal, SIGDANGER, which you get
   * when memory is exhausted.  since malloc() can overcommit, it's a good
   * idea to trap this one.
   */
  signal(SIGDANGER, memoryExhausted);
#endif

  dispcells = DisplayCells(Disp, Scrn);
  theCmap   = DefaultColormap(Disp, Scrn);

  if (argc < 2)
    usage(argv[0]);

  /* defaults and other initial settings.  some of these depend on what
   * our name was when invoked.
   */

  loadPathsAndExts();
  onroot= 0;
  verbose= 1;
  if (!strcmp(tail(argv[0]), "xview")) {
    onroot= 0;
    verbose= 1;
  }
  else if (!strcmp(tail(argv[0]), "xsetbg")) {
    onroot= 1;
    verbose= 0;
  }
  border= NULL;
  dname= NULL;
  fit= 0;
  fullscreen= 0;
  do_fork= 0;
  identify= 0;
  install= 0;
  private_cmap= 0;
  use_pixmap= 0;
  set_default= 0;
  dest_window= 0;
  user_geometry = NULL;
  winwidth= winheight= 0;
  visual_class= -1;

  imagecount= 0;
  for (a= 0; a < MAXIMAGES; a++) {
    images[a].name= NULL;
    images[a].atx= images[a].aty= 0;
    images[a].bright= 0;
    images[a].center= 0;
    images[a].clipx= images[a].clipy= 0;
    images[a].clipw= images[a].cliph= 0;
    images[a].colors= 0;
    images[a].delay= 0;
    images[a].dither= 0;
    images[a].gamma= 1.0;
    images[a].go_to= NULL;
    images[a].gray= 0;
    images[a].merge= 0;
    images[a].normalize= 0;
    images[a].rotate= 0;
    images[a].smooth= 0;
    images[a].xzoom= images[a].yzoom= 0;
    images[a].fg= images[a].bg= NULL;
  }
  for (a= 1; a < argc; a++) {
    switch (optionNumber(argv[a])) {
    case OPT_BADOPT:
      printf("%s: Bad option\n", argv[a]);
      usage(argv[0]);
      /* NOTREACHED */

    case OPT_NOTOPT:
      if (imagecount == MAXIMAGES)
	printf("%s: Too many images (ignoring)\n", argv[++a]);
      else {
	images[imagecount++].name= argv[a];
	if (imagecount < MAXIMAGES) {
	  images[imagecount].bright= global_bright;
	  images[imagecount].colors= global_colors;
	  images[imagecount].delay= global_delay;
	  images[imagecount].dither= global_dither;
	  images[imagecount].gamma= global_gamma;
	  images[imagecount].normalize= global_normalize;
	  images[imagecount].smooth= global_smooth;
	  images[imagecount].xzoom= global_xzoom;
	  images[imagecount].yzoom= global_yzoom;
	}
      }
      break;

    case OPT_SHORTOPT:
      printf("%s: Not enough characters to identify option\n", argv[a]);
      usage(argv[0]);
      /* NOTREACHED */

    /* process options global to everything
     */

    case OA_ONROOT:
      onroot= 1;
      fit= 1; /* assume -fit */
      break;

    case OA_BORDER:
      if (argv[++a])
	border= argv[a];
      break;

    case OA_DBUG:
      _Xdebug= True;
      break;

    case OA_DEFAULT:
      set_default= 1;
      break;

    case OA_DELAY:
      if (!argv[++a])
	  break;
      global_delay= atoi(argv[a]);
      if (global_delay <= 0) {
	  global_delay= 0;
	  printf("Bad argument to -delay\n");
	  break;
      }
      images[imagecount].delay= global_delay;
      break;

    case OA_DISPLAY:
      if (argv[++a])
	dname= argv[a];
      break;

    case OA_FIT:
      fit= 1;
      break;

    case OA_FORK:
      do_fork= 1;
      verbose= 0; /* background processes should be seen but not heard */
      break;

    case OA_FULLSCREEN:
      fullscreen= 1;
      break;

    case OA_GEOMETRY:
      if (argv[++a])
	user_geometry = argv[a];
      break;

    case OA_HELP:
      if (argv[++a])
	do {
	  help(argv[a++]);
	} while (argv[a]); 
      else
	help(NULL);
      exit(0);

    case OA_IDENTIFY:
      identify= 1;
      break;

    case OA_LIST:
      listImages();
      exit(0);

    case OA_INSTALL:
      install= 1;
      break;

    case OA_PATH:
      showPath();
      break;

    case OA_PIXMAP:
      use_pixmap= 1;
      break;

    case OA_PRIVATE:
      private_cmap= 1;
      break;

    case OA_QUIET:
      verbose= 0;
      break;

    case OA_SUPPORTED:
      supportedImageTypes();
      break;

    case OA_VERBOSE:
      verbose= 1;
      break;

    case OA_VER_NUM:
      version();
      break;

    case OA_VIEW:
      onroot= 0;
      break;

    case OA_VISUAL:
      if (argv[++a])
	visual_class= visualClassFromName(argv[a]);
      break;

    case OA_WINDOWID:
      if (!argv[++a])
	break;
      if (sscanf(argv[a], "0x%x", &dest_window) != 1) {
	printf("Bad argument to -windowid\n");
	usage(argv[0]);
	/* NOTREACHED */
      }
      onroot= 1; /* this means "on special root" */
      fit= 1; /* assume -fit */
      break;

    /* process options local to an image
     */

    case OA_AT:
      if (!argv[++a])
	break;
      if (sscanf(argv[a], "%d,%d",
		 &images[imagecount].atx, &images[imagecount].aty) != 2) {
	printf("Bad argument to -at\n");
	usage(argv[0]);
	/* NOTREACHED */
      }
      break;

    case OA_BACKGROUND:
      if (argv[++a])
	images[imagecount].bg= argv[a];
      break;

    case OA_BRIGHT:
      if (argv[++a]) {
	images[imagecount].bright= atoi(argv[a]);
	global_bright= images[imagecount].bright;
      }
      break;

    case OA_GAMMA:
      if (argv[++a]) {
	images[imagecount].gamma= atof(argv[a]);
	global_gamma= images[imagecount].gamma;
      }
      break;

    case OA_GOTO:
      if (argv[++a])
	images[imagecount].go_to= argv[a];
      break;

    case OA_GRAY:
      images[imagecount].gray= 1;
      break;

    case OA_CENTER:
      images[imagecount].center= 1;
      break;

    case OA_CLIP:
      if (!argv[++a])
	break;
      if (sscanf(argv[a], "%d,%d,%d,%d",
		 &images[imagecount].clipx, &images[imagecount].clipy,
		 &images[imagecount].clipw, &images[imagecount].cliph) != 4) {
	printf("Bad argument to -clip\n");
	usage(argv[0]);
	/* NOTREACHED */
      }
      break;

    case OA_COLORS:
      if (!argv[++a])
	break;
      images[imagecount].colors= atoi(argv[a]);
      if (images[imagecount].colors < 2) {
	printf("Argument to -colors is too low (ignored)\n");
	images[imagecount].colors= 0;
      }
      else if (images[imagecount].colors > 65536) {
	printf("Argument to -colors is too high (ignored)\n");
	images[imagecount].colors= 0;
      }
      global_colors= images[imagecount].colors;
      break;

    case OA_DITHER:
      images[imagecount].dither= 1;
      global_dither= 1;
      break;

    case OA_FOREGROUND:
      if (argv[++a])
	images[imagecount].fg= argv[a];
      break;

    case OA_HALFTONE:
      images[imagecount].dither= 2;
      global_dither= 2;
      break;

    case OA_IDELAY:
      if (!argv[++a])
	break;
      images[a].delay= atoi(argv[a]);
      if (images[a].delay <= 0) {
	printf("Bad argument to -idelay\n");
	images[a].delay= 0;
      }
      break;

    case OA_INVERT:
      images[imagecount].fg= "white";
      images[imagecount].bg= "black";
      break;

    case OA_MERGE:
      images[imagecount].merge= 1;
      break;

    case OA_NAME:
      if (imagecount == MAXIMAGES)
	printf("%s: Too many images (ignoring)\n", argv[++a]);
      else
	images[imagecount++].name= argv[++a];
      break;

    case OA_NEWOPTIONS:
      global_bright= 0;
      global_colors= 0;
      global_delay= 0;
      global_dither= 0;
      global_gamma= 1.0;
      global_normalize= 0;
      global_smooth= 0;
      global_xzoom= 0;
      global_yzoom= 0;
      break;

    case OA_NORMALIZE:
      images[imagecount].normalize= 1;
      global_normalize= images[imagecount].normalize;
      break;

    case OA_ROTATE:
      if (!argv[++a])
	break;
      images[imagecount].rotate = atoi(argv[a]);
      if ((images[imagecount].rotate % 90) != 0)
	{ printf("Argument to -rotate must be a multiple of 90 (ignored)\n");
	  images[imagecount].rotate = 0;
	}
      else 
	while (images[imagecount].rotate < 0)
	  images[imagecount].rotate += 360;
      break;

    case OA_SMOOTH:
      global_smooth++;
      images[imagecount].smooth= global_smooth;
      break;

    case OA_XZOOM:
      if (argv[++a]) {
	if (atoi(argv[a]) < 0) {
	  printf("Zoom argument must be positive (ignored).\n");
	  continue;
	}
	images[imagecount].xzoom= atoi(argv[a]);
	global_xzoom= images[imagecount].xzoom;
      }
      break;

    case OA_YZOOM:
      if (argv[++a]) {
	if (atoi(argv[a]) < 0) {
	  printf("Zoom argument must be positive (ignored).\n");
	  continue;
	}
	images[imagecount].yzoom= atoi(argv[a]);
	global_yzoom= images[imagecount].yzoom;
      }
      break;

    case OA_ZOOM:
      if (argv[++a]) {
	if (atoi(argv[a]) < 0) {
	  printf("Zoom argument must be positive (ignored).\n");
	  continue;
	}
	images[imagecount].xzoom= images[imagecount].yzoom= atoi(argv[a]);
	global_xzoom= global_yzoom= images[imagecount].xzoom;
      }
      break;

    default:

      /* this should not happen!
       */

      printf("%s: Internal error parsing arguments\n", argv[0]);
      exit(1);
    }
  }

  if (fit && (visual_class != -1)) {
    printf("-fit and -visual options are mutually exclusive (ignoring -visual)\n");
    visual_class= -1;
  }

  if (!imagecount && !set_default) /* NO-OP from here on */
    exit(0);

  if (identify) {                    /* identify the named image(s) */
    for (a= 0; a < imagecount; a++)
      identifyImage(images[a].name);
    exit(0);
  }

  /* start talking to the display
   */

  if (! (Disp= disp= XOpenDisplay(dname))) {
    printf("%s: Cannot open display\n", XDisplayName(dname));
    exit(1);
  }
  Scrn= scrn= DefaultScreen(disp);
  XSetErrorHandler(errorHandler);

  /* background ourselves if the user asked us to
   */

  if (do_fork)
    switch(fork()) {
    case -1:
      perror("fork");
      /* FALLTHRU */
    case 0:
      break;
    default:
      exit(0);
    }

  /* handle -default option.  this resets the colormap and loads the
   * default root weave.
   */

  if (set_default) {
    byte *old_data;

    dispimage= newBitImage(root_weave_width, root_weave_height);
    old_data= dispimage->data;
    dispimage->data= root_weave_bits;
    imageOnRoot(disp, scrn, dest_window, dispimage, 0);
    dispimage->data= old_data;
    freeImage(dispimage);
    if (!imagecount) /* all done */
      exit(0);
  }

  dispimage= NULL;

  if (onroot && (winwidth || winheight || images[0].center ||
      images[0].atx || images[0].aty || fullscreen)) {
    if (!winwidth)
	winwidth= DisplayWidth(disp, scrn);
    if (!winheight)
      winheight= DisplayHeight(disp, scrn);
    if (DefaultDepth(disp, scrn) == 1)
      dispimage= newBitImage(winwidth, winheight);
    else {
      dispimage= newRGBImage(winwidth, winheight, DefaultDepth(disp, scrn));
      dispimage->rgb.used= 1;
    }
    *(dispimage->rgb.red)= 65535;   /* default border value is white */
    *(dispimage->rgb.green)= 65535;
    *(dispimage->rgb.blue)= 65535;
    if (border) {
      XParseColor(disp, DefaultColormap(disp, scrn), border, &xcolor);
      *dispimage->rgb.red= xcolor.red;
      *dispimage->rgb.green= xcolor.green;
      *dispimage->rgb.blue= xcolor.blue;
    }

    /* bitmap needs both black and white
     */

    if (DefaultDepth(disp, scrn) == 1) {
	if (*(dispimage->rgb.red)) {
	    *(dispimage->rgb.red + 1)= 0;
	    *(dispimage->rgb.green + 1)= 0;
	    *(dispimage->rgb.blue + 1)= 0;
	}
	else {
	    *(dispimage->rgb.red + 1)= 65535;
	    *(dispimage->rgb.green + 1)= 65535;
	    *(dispimage->rgb.blue + 1)= 65535;
	}
    }
    fill(dispimage, 0, 0, winwidth, winheight, 0);
  }

  /* load in each named image
   */

  for (a= 0; a < imagecount; a++) {
    if (! (newimage= loadImage(images[a].name, verbose)))
      continue;
    if (!images[a].dither &&
	((dispimage && BITMAPP(dispimage)) || (DefaultDepth(disp, scrn) == 1)))
      images[a].dither= 1;

    /* if this is the first image and we're putting it on the root window
     * in fullscreen mode, set the zoom factors and
     * location to something reasonable.
     */

    if ((a == 0) && onroot && fullscreen &&
	(images[0].xzoom == 0.0) && (images[0].yzoom == 0.0) &&
	(images[0].atx == 0) && (images[0].aty == 0) &&
	(images[0].center == 0)) {
      XWindowAttributes wa;
      images[0].center= 1;
    
      if ((newimage->width > DisplayWidth(disp, scrn)) ||
	  (newimage->height > DisplayHeight(disp, scrn))) {
	images[0].xzoom= images[0].yzoom=
	  (newimage->width - DisplayWidth(disp, scrn) >
	   newimage->height - DisplayHeight(disp, scrn) ?
	   (float)DisplayWidth(disp, scrn) / (float)newimage->width * 100.0 :
	   (float)DisplayHeight(disp, scrn) / (float)newimage->height * 100.0);
      }
      else {
	images[0].xzoom= images[0].yzoom=
	  (DisplayWidth(disp, scrn) - newimage->width <
	   DisplayHeight(disp, scrn) - newimage->height ?
	   (float)DisplayWidth(disp, scrn) / (float)newimage->width * 100.0 :
	   (float)DisplayHeight(disp, scrn) / (float)newimage->height * 100.0);
      }
    }

    newimage= processImage(disp, scrn, newimage, &images[a], verbose);

    if (images[a].center && dispimage) {
      images[a].atx= (int)(dispimage->width - newimage->width) / 2;
      images[a].aty= (int)(dispimage->height - newimage->height) / 2;
    }
    if (dispimage) {
      if (! dispimage->title)
	dispimage->title= dupString(newimage->title);
      tmpimage= merge(dispimage, newimage, images[a].atx, images[a].aty,
		      verbose);
      if (dispimage != tmpimage) {
	freeImage(dispimage);
	dispimage= tmpimage;
      }
      freeImage(newimage);
    }
    else
      dispimage= newimage;

    imgObj[imgObjIdx].image = dispimage;

    printf("MADE imgObj[%d]  (%d,%d)\n", 
	   imgObjIdx-1, dispimage->width, dispimage->height);

    imgObj[imgObjIdx].ximageinfo = 
      imageInWindow(disp, scrn, dispimage, user_geometry,
		 fullscreen, install, private_cmap, fit, use_pixmap,
		 images[a].delay, visual_class, argc, argv, verbose);

    imgObjIdx++;

#ifdef NONO
    /* if next image is to be merged onto this one, do it now
     */

    if (onroot || ((a < imagecount) && (images[a + 1].merge)))
      continue;

    switch(imageInWindow(disp, scrn, dispimage, user_geometry,
			 fullscreen, install, private_cmap, fit, use_pixmap,
			 images[a].delay, visual_class, argc, argv, verbose)) {
    case '\0': /* window got nuked by someone */
      XCloseDisplay(disp);
      exit(1);
    case '\003':
    case 'q':  /* user quit */
      cleanUpWindow(disp);
      XCloseDisplay(disp);
      exit(0);
    case ' ':
    case 'n':  /* next image */
      if (images[a + 1].go_to) {
	int b;
	for (b= 0; b < imagecount; b++)
	  if (!strcmp(images[b].name, images[a + 1].go_to)) {
	    a= b - 1;
	    goto next_image;
	  }
	fprintf(stderr, "Target for -goto %s was not found\n",
		images[a + 1].go_to);
      next_image:
	;
      }
      break;
    case 'p':  /* previous image */
      if (a > 0)
	a -= 2;
      else
	a--;
      break;
    }
#endif
/*    freeImage(dispimage);
*/
    dispimage= NULL;

  }
  {
    int i;
    XSetWindowAttributes attrs;
    XEvent event;
    void processEvent();
    int ForegroundPixel, BackgroundPixel;

    ForegroundPixel = BlackPixel(disp, scrn);
    BackgroundPixel = WhitePixel(disp, scrn);
    gc = DefaultGC(disp, scrn);

    for (i = 0; i < imgObjIdx; i++) {
      attrs.background_pixel = BackgroundPixel;
      attrs.border_pixel = ForegroundPixel;
      attrs.event_mask = StructureNotifyMask | KeyPressMask |
			ButtonPressMask | ButtonReleaseMask | 
			ExposureMask;

      imgObj[i].w = XCreateWindow (disp, RootWindow(disp, scrn),
				   (int)10 + (i * 50), (int)10,
				   (int)imgObj[i].image->width,
				   (int)imgObj[i].image->height,
				   (int)1, /* border thickness */
				   CopyFromParent,
				   CopyFromParent,
				   CopyFromParent,
				   CWBackPixel | CWBorderPixel | CWEventMask,
				   &attrs);
      XMapRaised(disp, imgObj[i].w);
      XFlush(disp);

    }

    printf("Press right button to exit.\n");
    while(1){
      XNextEvent(disp, &event);
      processEvent(&event);
    }
  }

/*

  if (onroot)
    imageOnRoot(disp, scrn, dest_window, dispimage, verbose);
  XCloseDisplay(disp);
  exit(0);
*/
}

void processEvent(event)
     XEvent *event;
{
  int i, x,y;
  Window w = event->xany.window;

  printf("window=x%x\n", w);

  switch(event->type) {
  case ButtonPress :
    x = ((XButtonEvent *)event)->x;
    y = ((XButtonEvent *)event)->y;
    printf("x=%d, y=%d\n", x, y);

    switch(((XButtonPressedEvent *)event)->button) {
    case 1: /* left button */
      printf("left button press\n");
      for (i = 0; i < imgObjIdx; i++) {
	if (imgObj[i].w == w) {
	  printf("DESTROYING ImgObj[%d]\n", i);
	  freeXImage(imgObj[i].image, imgObj[i].ximageinfo);
	  freeImage(imgObj[i].image);
	  XDestroyWindow(Disp, imgObj[i].w);
	  imgObj[i].w = 0;
	  imgObj[i].image = 0;
	  imgObj[i].ximageinfo = 0;
	}
      }
      break;

    case 2: /* middle button */
      printf("middle button press\n");
      break;

    case 3: /* right button */
      printf("right button press: \n");
      break;
    }
    break;

  case Expose :
    {
      int x1, y1, x2, y2;

      printf("expose\n");

      for (i = 0; i < imgObjIdx; i++) {
	if (imgObj[i].w == w) {
	  printf("EXPOSURE FOR ImgObj[%d]\n", i);
	  XPutImage(Disp, w, gc, imgObj[i].ximageinfo->ximage, 0, 0, 
		    0, 0, imgObj[i].image->width, imgObj[i].image->height);
	  XFlush(Disp);
	}
      }
    }
  }
}

